File: npm/sharpearena/src/specHash.ts
/**
 * The tape-semantics spec hash this wrapper (and its committed `pkg/` wasm bundle)
 * was generated against — the value in
 * `crates/sharpearena/contract/attestation/spec-hash.json`.
 *
 * At module load, `index.ts` compares this pin to the hash the wasm engine itself
 * reports (`spec_hash()`, compiled in by the core crate's `build.rs` over the
 * tape-defining sources) and refuses to run on a mismatch. That makes "a wasm
 * bundle built from commit A driven by a wrapper from commit B" a named error
 * instead of a silently wrong number. Rebind this constant (and rebuild `pkg/`)
 * whenever the committed spec-hash record moves.
 */
export const SPEC_HASH = "5518afd039aa5317";

/**
 * Compare an engine-reported spec hash against **this wrapper's pin**, throwing the
 * named mismatch diagnosis on disagreement. `engineHash === undefined` means the
 * loaded engine predates the handshake entirely, the one frame of this exchange
 * that is decoded leniently, precisely so a stale surface is diagnosed by name
 * rather than dropped as malformed.
 *
 * The pin is deliberately not a parameter. An exported `check(a, b)` reads as
 * "compare these two", not "compare against the pin", and `check(h, h)` passes for
 * any `h`. This is the only spec-hash entry point the package exposes, so a caller
 * cannot supply the value it is being checked against.
 */
export function checkSpecHash(engineHash: string | undefined): void {
  compareSpecHash(engineHash, SPEC_HASH);
}

/**
 * The two-sided comparison behind {@link checkSpecHash}. Internal: it is not
 * re-exported from the package entry point, because a caller-supplied `wrapperHash`
 * turns the handshake into a tautology. The stale-surface regression drives this
 * function directly, which is why it is named rather than inlined.
 */
export function compareSpecHash(
  engineHash: string | undefined,
  wrapperHash: string,
): void {
  if (engineHash === wrapperHash) return;
  if (engineHash === undefined) {
    throw new Error(
      `SpecHashMismatch: the loaded wasm engine predates the SPEC_HASH handshake ` +
        `(no spec_hash export); wrapper built against 0x${wrapperHash}. ` +
        `Rebuild pkg/ from the same commit as this wrapper.`,
    );
  }
  throw new Error(
    `SpecHashMismatch: engine spec 0x${engineHash}, wrapper built against ` +
      `0x${wrapperHash}. The committed wasm bundle and this wrapper come from ` +
      `different tape-semantics revisions; rebuild pkg/ from the same commit.`,
  );
}

File: crates/sharpearena-py/python/sharpearena/_spec_hash.py
"""The cross-surface SPEC_HASH handshake for the Python package.

The Rust core's ``build.rs`` fingerprints the tape-semantics-defining sources
(FNV-1a/64 over CRLF-normalized bytes, seeded with a manual epoch) and compiles the
value into the engine; the pyo3 extension reports it as ``spec_hash()``. This module
pins the hash the pure-Python package was generated against
(``crates/sharpearena/contract/attestation/spec-hash.json``) and
:func:`check_spec_hash` refuses at package import when the compiled extension
disagrees, so a stale ``.pyd``/``.so`` in a dev venv driven by newer Python sources
(or a stale wheel layered under newer sources) is a named error, not a silently
wrong number.

The engine's ``spec_hash`` attribute is the ONE thing read leniently
(:func:`engine_spec_hash` returns ``None`` instead of raising when the extension
predates the handshake), precisely so a stale surface is diagnosed by name rather
than dying with an ``AttributeError`` that says nothing about versions.
"""

from __future__ import annotations

from typing import Optional

# The pin. Rebind together with contract/attestation/spec-hash.json (a Rust test and
# tests/test_spec_hash.py keep the three copies bound).
EXPECTED_SPEC_HASH = "2d912913eedd333e"


class SpecHashMismatch(RuntimeError):
    """The compiled engine and the pure-Python package disagree on tape semantics."""


def engine_spec_hash(engine: object) -> Optional[str]:
    """The engine's reported spec hash, or ``None`` when it predates the handshake."""
    fn = getattr(engine, "spec_hash", None)
    return None if fn is None else str(fn())


def check_spec_hash(engine_hash: Optional[str], expected: str = EXPECTED_SPEC_HASH) -> None:
    """Refuse a stale engine/wrapper pairing with a named diagnosis."""
    if engine_hash == expected:
        return
    if engine_hash is None:
        raise SpecHashMismatch(
            "the compiled sharpearena_py extension predates the SPEC_HASH handshake "
            f"(no spec_hash export); wrapper built against 0x{expected}. "
            "Rebuild the extension (maturin develop) from the same commit as the "
            "Python sources."
        )
    raise SpecHashMismatch(
        f"engine spec 0x{engine_hash}, wrapper built against 0x{expected}. "
        "The compiled sharpearena_py extension and the Python package come from "
        "different tape-semantics revisions; rebuild the extension (maturin develop) "
        "from the same commit as the Python sources."
    )
